src/Pos/Client/Txp/Balances.hs

```diff
@@ -68,17 +69,28 @@ runBalancesRedirect = coerce
 
 instance
     ( MonadDBRead m
+    , MonadRealDB m
     , WebWalletModeDB m
-    ,  MonadMask m
+    , MonadMask m
     , WithLogger m
     , MonadTxpMem ext m, t ~ IdentityT) =>
         MonadBalances (Ether.TaggedTrans BalancesRedirectTag t m)
   where
-    getOwnUtxos addr = do
+    getOwnUtxos addrs = do
+        let isRedeem (RedeemAddress _) = True
+            isRedeem _                 = False
+            redeemAddrs = filter isRedeem addrs
+            commonAddrs = filter (not . isRedeem) addrs
+
         updates <- getUtxoModifier
-        locUtxo <- MM.modifyMap updates <$> WS.getWalletUtxo
-        let addrsSet = HS.fromList $ AddressIA <$> addr
-        pure $ M.filter (`addrBelongsToSet` addrsSet) locUtxo
+        commonUtxo <- if null commonAddrs then pure mempty
+                      else WS.getWalletUtxo
+        redeemUtxo <- if null redeemAddrs then pure mempty
```

Comment:
you can use `genesisUtxoM` to avoid iterating over whole Utxo, right?
but I think we can do it in the next PR
@flyingleafe